### ISSUE TITLE
HEAD requests return 500 Server Error

### ISSUE THREAD

**Issue by fbonzon** [graphhopper/graphhopper#677]
If you query the API for a GPX file with a HEAD HTTP request, the server will always fail with HTTP error 500 Server Error, as far as I could test.

E.g.:

```
curl -I 'https://graphhopper.com/api/1/route?point=52.517037%2C13.38886&point=52.515437%2C13.362293&type=gpx&key=xxxxxxxxxxxxx&gpx.route=false&gpx.track=true&gpx.waypoints=true&locale=en-US&vehicle=bike&weighting=fastest&elevation=true&latlng=46.945705%2C7.470643'
```

Returns:

```
HTTP/1.1 500 Server Error
Server: nginx
Date: Sun, 06 Mar 2016 04:09:32 GMT
Content-Type: text/html;charset=iso-8859-1
Content-Length: 381
Connection: keep-alive
Cache-Control: must-revalidate,no-cache,no-store
```

But regular GET request to the same URL succeeds:

```
curl -i 'https://graphhopper.com/api/1/route?point=52.517037%2C13.38886&point=52.515437%2C13.362293&type=gpx&key=xxxxxxxxxxxxx&gpx.route=false&gpx.track=true&gpx.waypoints=true&locale=en-US&vehicle=bike&weighting=fastest&elevation=true&latlng=46.945705%2C7.470643'
```

Returns:

```
HTTP/1.1 200 OK
Server: nginx
Date: Sun, 06 Mar 2016 04:10:02 GMT
Content-Type: application/xml
Content-Length: 2173
Connection: keep-alive
Access-Control-Allow-Origin: *
X-RateLimit-Limit: 500
X-RateLimit-Remaining: 261
X-RateLimit-Reset: 38040
Content-Disposition: attachment;filename=GraphHopper.gpx
X-RateLimit-Credits: 1
Strict-Transport-Security: max-age=31536000; includeSubDomains;

<?xml version="1.0" encoding="UTF-8" standalone="no" ?><gpx xmlns="http://www.topografix.com/GPX/1/1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" creator="Graphhopper version 0.7" version="1.1" xmlns:gh="https://graphhopper.com/public/schema/gpx/1.1">
<metadata><copyright author="OpenStreetMap contributors"/><link href="http://graphhopper.com"><text>GraphHopper GPX</text></link><time>2016-03-06T04:10:02Z</time></metadata>
<wpt lat="52.517033" lon="13.3888"> <name>Friedrichstraße</name></wpt>
<wpt lat="52.515344" lon="13.362309"> <name>Finish!</name></wpt>
<trk><name>GraphHopper Track</name><trkseg>
<trkpt lat="52.517033" lon="13.3888"><ele>43.2</ele><time>2016-03-06T04:10:02Z</time></trkpt>
(...)
<trkpt lat="52.515344" lon="13.362309"><ele>37.7</ele><time>2016-03-06T04:17:11Z</time></trkpt>
</trkseg>
</trk>
</gpx>
```

For background, I see 2 use cases for issuing a HEAD request:
- (my use case) Know the file size prior to downloading the entire file, in case you want to enforce a maximum size.
- Look at the `X-RateLimit-Remaining: 262` header to check your remaining API quota.

As the HTTP spec suggests, ideally a HEAD request would return the same headers as a GET request, without the body.

**Comment by karussell** [comment]
This would be only achievable via some caching mechanism which I would like to avoid. Or how do you see this could be implemented without a cache?

> Look at the X-RateLimit-Remaining: 262 header to check your remaining API quota.

That is related to the Directions API, issues should go [here](https://github.com/graphhopper/directions-api/issues). Instead of a head one could introduce a new API request

**Comment by fbonzon** [comment]
Sure, returning a more appropriate HTTP code like "405 Method Not Allowed" is an acceptable and easy solution. To mean that the result is intentional, and that the server does not implement HEAD request. Code 500 implies a server problem, which is apparently not true here.

If you decide to actually implement HEAD requests: Yes, using a short-lived cache is the ideal solution. A subsequent GET request for the same URL would come faster, and then clear the cache. And you would probably bill the credits only once, for both HEAD and requests.

Another solution is simply to not use any cache, and bill credits as a GET request. If subsequent GET request for the same URL comes later, credits are billed again. It would be interesting to know how other API deal with this HEAD requests / billing issue.

I don't have the need for an API endpoint to check remaining API quota, so I'll won't create new issue there. It was a side note about the usefulness of HEAD requests.

**Comment by karussell** [comment]
> And you would probably bill the credits only once, for both HEAD and requests.

Yes, exactly that is the reason we would need a cache. Hmmh, but if someone comes up with a real use case for HEAD requests we would need something like this and as you are saying via a 'short-lived' cache. Or if the cache is necessary for other things ...

**Comment by fbonzon** [comment]
Yes. Change the returned HTTP code now, and add the implementation of HEAD request in the wish list.
